[Filters] Implement identity filters

shallawa opened the pull request:
#### 4ac8e180ed02f9af4ed1cc1742b6365bea860c7d
<pre>
[Filters] Implement identity filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=242359">https://bugs.webkit.org/show_bug.cgi?id=242359</a>
rdar://96907495

Reviewed by Simon Fraser.

We should draw the element as if there was no filter applied if the whole filter
chain is no-operation. Identity filters are common scenario after transitioning
the filter parameters to identity values. This is a performance and  memory
enhancement for filers.

* LayoutTests/css3/filters/noop-filters-expected.html: Added.
* LayoutTests/css3/filters/noop-filters.html: Added.
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::isNoopBlurEffect):
(WebCore::isNoopComponentTransferFilterffect):
(WebCore::isNoopDropShadowEffect):
(WebCore::isNoopColorMatrixFilterEffect):
(WebCore::isNoopReferenceFilter):
(WebCore::CSSFilter::isNoop):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::isNoop):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::postApplyResource):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h:
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::isNoop const):
* Source/WebCore/svg/SVGFEDropShadowElement.h:
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
(WebCore::SVGFEGaussianBlurElement::isNoop const):
* Source/WebCore/svg/SVGFEGaussianBlurElement.h:
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(WebCore::SVGFEMorphologyElement::isNoop const):
* Source/WebCore/svg/SVGFEMorphologyElement.h:
* Source/WebCore/svg/SVGFEOffsetElement.cpp:
(WebCore::SVGFEOffsetElement::isNoop const):
* Source/WebCore/svg/SVGFEOffsetElement.h:
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(WebCore::SVGFilterPrimitiveStandardAttributes::isNoop const):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::isNoop):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:

Canonical link: <a href="https://commits.webkit.org/253098@main">https://commits.webkit.org/253098@main</a>
</pre>

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilter.h:
```diff
@@ -39,6 +39,7 @@ class SVGFilter final : public Filter {
     static RefPtr<SVGFilter> create(SVGFilterElement&, RenderingMode, const FloatSize& filterScale, ClipOperation, const FloatRect& filterRegion, const FloatRect& targetBoundingBox, const GraphicsContext& destinationContext);
     WEBCORE_EXPORT static RefPtr<SVGFilter> create(const FloatRect& targetBoundingBox, SVGUnitTypes::SVGUnitType primitiveUnits, SVGFilterExpression&&);
 
+    static bool isNoop(SVGFilterElement&);
```
Maybe `isIdentity()` would be easier to read (replace all the "Noop" with "isIdentity").

smfr commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -75,6 +75,11 @@ CSSFilter::CSSFilter(Vector<Ref<FilterFunction>>&& functions)
 {
 }
 
+static bool isIdentityBlurEffect(const BlurFilterOperation& blurOperation)
```
You could either rely on C++ overloading and have all these static functions be called `isIdentityEffect()`, or you could add a `virtual bool isIdentity(std::optional<float> noopAmount) const` to the base class.

shallawa committed:
[Filters] Implement identity filters
https://bugs.webkit.org/show_bug.cgi?id=242359
rdar://96907495

Reviewed by Simon Fraser.

We should draw the element as if there was no filter applied if the whole filter
chain is no-operation. Identity filters are common scenario after transitioning
the filter parameters to identity values. This is a performance and  memory
enhancement for filers.

* LayoutTests/css3/filters/noop-filters-expected.html: Added.
* LayoutTests/css3/filters/noop-filters.html: Added.
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::isNoopBlurEffect):
(WebCore::isNoopComponentTransferFilterffect):
(WebCore::isNoopDropShadowEffect):
(WebCore::isNoopColorMatrixFilterEffect):
(WebCore::isNoopReferenceFilter):
(WebCore::CSSFilter::isNoop):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintsWithFilters const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::isNoop):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::postApplyResource):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.h:
* Source/WebCore/svg/SVGFEDropShadowElement.cpp:
(WebCore::SVGFEDropShadowElement::isNoop const):
* Source/WebCore/svg/SVGFEDropShadowElement.h:
* Source/WebCore/svg/SVGFEGaussianBlurElement.cpp:
(WebCore::SVGFEGaussianBlurElement::isNoop const):
* Source/WebCore/svg/SVGFEGaussianBlurElement.h:
* Source/WebCore/svg/SVGFEMorphologyElement.cpp:
(WebCore::SVGFEMorphologyElement::isNoop const):
* Source/WebCore/svg/SVGFEMorphologyElement.h:
* Source/WebCore/svg/SVGFEOffsetElement.cpp:
(WebCore::SVGFEOffsetElement::isNoop const):
* Source/WebCore/svg/SVGFEOffsetElement.h:
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(WebCore::SVGFilterPrimitiveStandardAttributes::isNoop const):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::isNoop):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:

Canonical link: https://commits.webkit.org/253098@main
